viz/RobotModel.cpp

```diff
@@ -399,21 +619,172 @@ osg::ref_ptr<osg::Node> RobotModel::makeOsg( boost::shared_ptr<urdf::ModelInterf
+    SdfElementPtrMap links = loadSdfModelLinks(sdf_model);
+
+    KDL::Tree tree;
+    kdl_parser::treeFromSdfModel(sdf_model, tree);
+
+    KDL::SegmentMap::const_iterator root = tree.getRootSegment();
+
+    std::vector<KDL::SegmentMap::const_iterator> segment_buffer;
+    segment_buffer.insert(segment_buffer.end(), root->second.children.begin(), root->second.children.end());
+
+    std::vector<osg::Node*> hook_buffer;
+    for(uint i = 0; i < root->second.children.size(); i++){
+        hook_buffer.push_back(original_root_);
+    }
+
+    original_root_name_ = root->second.segment.getName();
+
+    osg::Node* hook = 0;
+    while (!segment_buffer.empty()){
+
+        KDL::SegmentMap::const_iterator it = segment_buffer.back();
+        KDL::SegmentMap::const_iterator parent_it = it->second.parent;
+
+        segment_buffer.pop_back();
+        KDL::Segment kdl_segment = it->second.segment;
+        KDL::Segment kdl_parent_segment = parent_it->second.segment;
+
+        if (it->second.children.size() > 0){
+            segment_buffer.insert(segment_buffer.end(), it->second.children.begin(), it->second.children.end());
+        }
+
+        std::map<std::string, sdf::ElementPtr>::iterator link_itr = links.find(kdl_segment.getName());
+        std::map<std::string, sdf::ElementPtr>::iterator link_parent_itr = links.find(kdl_parent_segment.getName());
+
+        if (link_itr != links.end()){
+            sdf::ElementPtr sdf_link = link_itr->second;
+            sdf::ElementPtr sdf_parent_link;
+
+            if (link_parent_itr != links.end()) {
+                sdf_parent_link = link_parent_itr->second;
+            }
+
+            //create osg link
+            hook = hook_buffer.back();
+            hook_buffer.pop_back();
+
+            osg::Node* new_hook = makeOsg2(kdl_segment, sdf_link, sdf_parent_link, hook->asGroup());
+
+            //Append hooks
+            for(uint i = 0; i < it->second.children.size(); i++){
+                hook_buffer.push_back(new_hook);
+            }
+
+            std::cout << "segment_name: " << kdl_segment.getName() << std::endl;
+
+            segmentNames_.push_back(kdl_segment.getName());
+            if(kdl_segment.getJoint().getType() != KDL::Joint::None)
+                jointNames_.push_back(kdl_segment.getJoint().getName());
+        }
+
+    }
+
+    /**
+     * in the SDF files it is not necessary one joint per two links
+     * in the SDF files it is possible to create a file with multiple links without joints
+     * original_root_ is not a segment, is only used to store the segments
+     * if the links in the SDF doesn't have a joint, then this segment will be a direct child of original_root_
+     * if all links doen't have a joint, then all segments will be children direct of original_root_
+     */
+    root_->addChild(original_root_);
+    return root_;
+}
+
 osg::ref_ptr<osg::Node> RobotModel::load(QString path){
```

Comment:
Allow the user to specify the loading method, to override auto-detection. Auto-detection will always fail one day or another ...